pkg/skaffold/runner/load_images.go

```diff
@@ -42,7 +42,7 @@ func (r *SkaffoldRunner) loadImagesInKindNodes(ctx context.Context, out io.Write
 func (r *SkaffoldRunner) loadImagesInK3dNodes(ctx context.Context, out io.Writer, k3dCluster string, artifacts []build.Artifact) error {
 	color.Default.Fprintln(out, "Loading images into k3d cluster nodes...")
 	return r.loadImages(ctx, out, artifacts, func(tag string) *exec.Cmd {
-		return exec.CommandContext(ctx, "k3d", "load", "image", "--cluster", k3dCluster, tag)
+		return exec.CommandContext(ctx, "k3d", "image", "import", "--cluster", k3dCluster, tag)
```

Comment:
Thanks for your contribution. So as to not break older versions can you add a conditional check on the `k3d` version and corresponding commands. Maybe create `load_images_k3d.go`  file.

Reply:
Sure, check k3d version and use corresponding cli is a good idea.

What prevent me from doing so is I'm not sure how to write test, which requires mock different k3d version result.

And collecting those different k3d version result need some efforts.

I will look into the issues later.

Thanks for your review.

Reply:
Hi @wenzong — take a look at the [helm deployer](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/deploy/helm.go) and its tests where we had to do something similar.

Reply:
> Hi @wenzong — take a look at the [helm deployer](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/deploy/helm.go) and its tests where we had to do something similar.

👍  Thanks a lot, that's very helpful. Will look into it.

Reply:
I've been looking at this and it seems k3d in it's history had a variety of version strings:

- `dev` https://github.com/rancher/k3d/commit/29fced4ef78bfecc7e1168fccfb88588a0145caa#diff-22d6d38cd847736f6c99f3133279067cR9
- [adding version command](https://github.com/rancher/k3d/commit/6b148ae7559ed5561dcf16328af68826da221745#diff-22d6d38cd847736f6c99f3133279067c)
- semver: https://github.com/rancher/k3d/commit/52b23adec9cb8c5317a278c35a015e2fb42e9736#diff-22d6d38cd847736f6c99f3133279067cR29
- `v3-dev`: https://github.com/rancher/k3d/commit/db03084e5d84177e745260dcf9816dfda167ad63#diff-22d6d38cd847736f6c99f3133279067cR39

Especially when running a locally built `v3-dev`, that shouldn't break skaffold support. So I figure it's best to ask for forgiveness rather than permission: and gracefully catch the error.

Reply:
## k3d-darwin-amd64.v1.0.0

### version:

```
No help topic for 'version'
```

### import-images:

```
No help topic for 'import-images'
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.1.0

### version:

```
No help topic for 'version'
```

### import-images:

```
No help topic for 'import-images'
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.3.4

### version:

```
No help topic for 'version'
```

### import-images:

```
NAME:
   k3d-darwin-amd64.v1.3.4 import-images - Import a comma- or space-separated list of container images from your local docker daemon into the cluster
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.4.0

### version:

```
k3d version v1.4.0
k3s version v1.0.1
```

### import-images:

```
NAME:
   k3d-darwin-amd64.v1.4.0 import-images - Import a comma- or space-separated list of container images from your local docker daemon into the cluster
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.5.0

### version:

```
k3d version v1.5.0
k3s version v1.17.0-k3s.1
```

### import-images:

```
NAME:
   k3d-darwin-amd64.v1.5.0 import-images - Import a comma- or space-separated list of container images from your local docker daemon into the cluster
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.6.0

### version:

```
k3d version v1.6.0
k3s version v1.17.2-k3s1
```

### import-images:

```
NAME:
   k3d-darwin-amd64.v1.6.0 import-images - Import a comma- or space-separated list of container images from your local docker daemon into the cluster
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v1.7.0

### version:

```
k3d version v1.7.0
k3s version v1.17.3-k3s1
```

### import-images:

```
NAME:
   k3d-darwin-amd64.v1.7.0 import-images - Import a comma- or space-separated list of container images from your local docker daemon into the cluster
```

### load image

```
No help topic for 'load'
```

### image import:

```
No help topic for 'image'
```

----

## k3d-darwin-amd64.v3.0.0-beta.1

### version:

```
k3d version v3.0.0-beta.1
```

### import-images:

```
Error: unknown command "import-images" for "k3d"
Run 'k3d --help' for usage.
```

### load image

```
Load an image from docker into a k3d cluster.

```

### image import:

```
Error: unknown command "image" for "k3d"
Run 'k3d --help' for usage.
```

----

## k3d-darwin-amd64.v3.0.0-rc.6

### version:

```
k3d version v3.0.0-rc.6
k3s version v1.18.4-k3s1 (default)
```

### import-images:

```
Error: unknown command "import-images" for "k3d"
Run 'k3d --help' for usage.
```

### load image

```
Load an image from docker into a k3d cluster.

```

### image import:

```
Error: unknown command "image" for "k3d"
Run 'k3d --help' for usage.
```

----

## k3d-darwin-amd64.v3.0.0-rc.7

### version:

```
k3d version v3.0.0-rc.7
k3s version v1.18.4-k3s1 (default)
```

### import-images:

```
Error: unknown command "import-images" for "k3d"
Run 'k3d --help' for usage.
```

### load image

```
Error: unknown command "load" for "k3d"
Run 'k3d --help' for usage.
```

### image import:

```
Import image(s) from docker into k3d cluster(s).

```

----

## k3d-darwin-amd64.v3.0.0

### version:

```
k3d version v3.0.0
k3s version latest (default)
```

### import-images:

```
Error: unknown command "import-images" for "k3d"
Run 'k3d --help' for usage.
```

### load image

```
Error: unknown command "load" for "k3d"
Run 'k3d --help' for usage.
```

### image import:

```
Import image(s) from docker into k3d cluster(s).

```

Reply:
Well, TL;DR skaffold never support k3d v1+, and I am wondering whether we still have to support v3.0.0.alpha ~ v3.0.0.rc6?

------

## v1

```
k3d import-images
```

## v3.0.0.alpha ~ v3.0.0.rc6

```
k3d load image
```

## v3.0.0.rc7 ~ v3.0.0

```
k3d image import
```

------